src/rust/common/src/ssl.rs

```diff
@@ -2,35 +2,71 @@
 // Licensed under the Apache License, Version 2.0
 // http://www.apache.org/licenses/LICENSE-2.0
 
+use boring::x509::X509;
 use boring::ssl::*;
 use config::TlsConfig;
 use std::io::{Error, ErrorKind};
 
 pub fn ssl_context(config: &TlsConfig) -> Result<Option<SslContext>, std::io::Error> {
     let mut builder = SslAcceptor::mozilla_intermediate_v5(SslMethod::tls_server())?;
 
-    if let Some(f) = config.certificate_chain() {
+    // load the private key
+    if let Some(f) = config.private_key() {
         builder
-            .set_ca_file(f)
-            .map_err(|_| Error::new(ErrorKind::Other, "bad certificate chain"))?;
+            .set_private_key_file(f, SslFiletype::PEM)
+            .map_err(|_| Error::new(ErrorKind::Other, "bad private key"))?;
     } else {
         return Ok(None);
     }
 
-    if let Some(f) = config.certificate() {
+    // load the ca file
+    if let Some(f) = config.ca_file() {
         builder
-            .set_certificate_file(f, SslFiletype::PEM)
-            .map_err(|_| Error::new(ErrorKind::Other, "bad certificate"))?;
+            .set_ca_file(f)
+            .map_err(|_| Error::new(ErrorKind::Other, "bad ca file"))?;
     } else {
         return Ok(None);
     }
 
-    if let Some(f) = config.private_key() {
-        builder
-            .set_private_key_file(f, SslFiletype::PEM)
-            .map_err(|_| Error::new(ErrorKind::Other, "bad private key"))?;
+    let chain = config.certificate_chain().is_some();
+    let cert = config.certificate().is_some();
+
+    if chain && !cert {
+        // assume we have a complete chain: leaf + intermediates + root in one file
+        if let Some(f) = config.certificate_chain() {
+            builder
+                .set_certificate_chain_file(f)
+                .map_err(|_| Error::new(ErrorKind::Other, "bad certificate chain"))?;
+        } else {
+            return Ok(None);
```

Comment:
when would this branch be taken (given `chain` is `true`)?

Reply:
Good point. We can just unwrap inside these branches